packages/expo-constants/ios/EXConstants/EXConstantsService+InstallationId.m

```diff
@@ -0,0 +1,105 @@
+// Copyright 2015-present 650 Industries. All rights reserved.
+
+#import <EXConstants/EXConstantsService+InstallationId.h>
+
+static NSString * const kEXDeviceInstallUUIDKey = @"EXDeviceInstallUUIDKey";
+
+@implementation EXConstantsService (InstallationId)
+
+- (NSString *)installationId
+{
+  NSString *installationId = [self fetchInstallationId];
+  if (installationId) {
+    return installationId;
+  }
+  
+  installationId = [[NSUUID UUID] UUIDString];
+  [self setInstallationId:installationId];
+  return installationId;
+}
+
+- (nullable NSString *)fetchInstallationId
+{
+  NSString *installationId;
+  CFTypeRef foundDict = NULL;
+  
+  if (SecItemCopyMatching((__bridge CFDictionaryRef)[self installationIdGetQuery], &foundDict) == noErr) {
+    NSData *result = (__bridge_transfer NSData *)foundDict;
+    NSString *value = [[NSString alloc] initWithData:result
+                                            encoding:NSUTF8StringEncoding];
+    // `initWithUUIDString` returns nil if string is not a valid UUID
+    if ([[NSUUID alloc] initWithUUIDString:value]) {
+      installationId = value;
+    }
+  }
+  
+  if (installationId) {
+    return installationId;
+  }
+  
+  NSString *legacyUUID = [[NSUserDefaults standardUserDefaults] stringForKey:kEXDeviceInstallUUIDKey];
+  if (legacyUUID) {
+    installationId = legacyUUID;
+    
+    NSError *error = [self setInstallationId:installationId];
+    if (error) {
+      NSLog(@"Could not migrate device install UUID from legacy storage: %@", error.description);
+    } else {
+      // We only remove the value from old storage once it's set and saved in the new storage.
+      [[NSUserDefaults standardUserDefaults] removeObjectForKey:kEXDeviceInstallUUIDKey];
+    }
+  }
+  
+  return installationId;
+}
+
+- (nullable NSError *)setInstallationId:(NSString *)installationId
```

Comment:
cocoa convention is to return a `BOOL` and accept `NSError **` (this also helps when translating to Swift)